Kernel/Time/TimeManagement.cpp

```diff
@@ -0,0 +1,249 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <Kernel/ACPI/ACPIParser.h>
+#include <Kernel/KParams.h>
+#include <Kernel/Scheduler.h>
+#include <Kernel/Time/HPET.h>
+#include <Kernel/Time/HPETComparator.h>
+#include <Kernel/Time/PIT.h>
+#include <Kernel/Time/RTC.h>
+#include <Kernel/Time/TimeManagement.h>
+#include <Kernel/VM/MemoryManager.h>
+
+//#define TIME_DEBUG
+
+namespace Kernel {
+
+static TimeManagement* s_time_management;
+
+bool TimeManagement::initialized()
+{
+    return s_time_management != nullptr;
+}
+
+bool TimeManagement::is_system_timer(const HardwareTimer& timer)
+{
+    return &timer == m_system_timer.ptr();
+}
+
+void TimeManagement::set_epoch_time(time_t value)
+{
+    InterruptDisabler disabler;
+    m_epoch_time = value;
+}
+
+time_t TimeManagement::epoch_time() const
+{
+    return m_epoch_time;
+}
+
+void TimeManagement::initialize(bool probe_non_legacy_hardware_timers)
+{
+    ASSERT(!TimeManagement::initialized());
+    s_time_management = new TimeManagement(probe_non_legacy_hardware_timers);
+}
+time_t TimeManagement::seconds_since_boot() const
+{
+    return m_seconds_since_boot;
+}
+time_t TimeManagement::ticks_per_second() const
+{
+    return m_system_timer->ticks_per_second();
+}
+
+time_t TimeManagement::ticks_this_second() const
+{
+    return m_ticks_this_second;
+}
+
+time_t TimeManagement::boot_time() const
+{
+    return RTC::boot_time();
+}
+
+void TimeManagement::stale_function(const RegisterState&)
+{
+}
+
+TimeManagement::TimeManagement(bool probe_non_legacy_hardware_timers)
+{
+    if (ACPI::Parser::the().is_operable()) {
+        if (!ACPI::Parser::the().x86_specific_flags().cmos_rtc_not_present) {
+            RTC::initialize();
+            m_epoch_time += boot_time();
+        } else {
+            klog() << "ACPI: RTC CMOS Not present";
+        }
+    } else {
+        // We just assume that we can access RTC CMOS, if ACPI isn't usable.
+        RTC::initialize();
+        m_epoch_time += boot_time();
+    }
+    if (probe_non_legacy_hardware_timers) {
+        if (!probe_and_set_non_legacy_hardware_timers())
+            if (!probe_and_set_legacy_hardware_timers())
+                ASSERT_NOT_REACHED();
+        return;
+    }
+    ASSERT(probe_and_set_legacy_hardware_timers());
+}
+
+Vector<size_t> TimeManagement::scan_and_set_periodic_timers()
```

Comment:
This name is a bit confusing; "setting a timer" sounds to me like it's starting a countdown.
What we're really doing here is something more like `initialize_periodic_timers()`, right?